library/core/src/str/validations.rs

```diff
@@ -112,162 +112,290 @@ where
-    1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, // 6
-    1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, // 7
-    0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, // 8
-    0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, // 9
-    0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, // A
-    0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, // B
-    0, 0, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, // C
-    2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, // D
-    3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, // E
-    4, 4, 4, 4, 4, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, // F
-];
+#[inline]
+const fn validate_non_acii_bytes(buf: &[u8], mut curr: usize) -> Result<usize, Utf8Error> {
+    const fn subarray<const N: usize>(buf: &[u8], idx: usize) -> Option<[u8; N]> {
+        if buf.len() - idx < N {
+            return None;
+        }
+
+        // SAFETY: checked in previous condition
+        Some(unsafe { *(buf.as_ptr().add(idx) as *const [u8; N]) })
+    }
+
+    let prev = curr;
+    macro_rules! err {
+        ($error_len: expr) => {
+            return Err(Utf8Error { valid_up_to: prev, error_len: $error_len })
+        };
+    }
+
+    let b0 = buf[curr];
+    match utf8_char_width(b0) {
+        2 => {
+            let Some([_, b1]) = subarray(buf, curr) else {
+                err!(None);
+            };
+
+            if b1 as i8 >= -64 {
+                err!(Some(1));
+            }
+
+            curr += 2;
+        }
+        3 => {
+            let Some([_, b1, b2]) = subarray(buf, curr) else {
+                err!(None);
+            };
+
+            match (b0, b1) {
+                (0xE0, 0xA0..=0xBF)
+                | (0xE1..=0xEC, 0x80..=0xBF)
+                | (0xED, 0x80..=0x9F)
+                | (0xEE..=0xEF, 0x80..=0xBF) => {}
+                _ => err!(Some(1)),
+            }
+
+            if b2 as i8 >= -64 {
+                err!(Some(2));
+            }
+
+            curr += 3;
+        }
+        4 => {
+            let Some([_, b1, b2, b3]) = subarray(buf, curr) else {
+                err!(None);
+            };
+
+            match (b0, b1) {
+                (0xF0, 0x90..=0xBF) | (0xF1..=0xF3, 0x80..=0xBF) | (0xF4, 0x80..=0x8F) => {}
+                _ => err!(Some(1)),
+            }
+
+            if b2 as i8 >= -64 {
+                err!(Some(2));
+            }
+
+            if b3 as i8 >= -64 {
+                err!(Some(3));
+            }
+
+            curr += 4;
+        }
+        _ => err!(Some(1)),
+    }
+
+    Ok(curr)
+}
+
+/// Returns `true` if any one block is not a valid ASCII byte.
+#[inline(always)]
+const fn has_non_ascii_byte<const N: usize>(block: &[usize; N]) -> bool {
+    let mut vector = [0; N];
+
+    let mut i = 0;
+    while i < N {
+        vector[i] = block[i] & NONASCII_MASK;
+        i += 1;
+    }
+
+    i = 0;
+    while i < N {
+        if vector[i] > 0 {
+            return true;
+        }
+        i += 1;
+    }
+
+    false
```

Comment:
```suggestion
    let mut i = 0;
    while i < N {
        if block[i] & NONASCII_MASK > 0 {
            return true;
        }
        i += 1;
    }

    false
```
Just curious if doing this in one block is sufficent or if I missed something.